src/metabase/pulse/markdown.clj

```diff
@@ -0,0 +1,305 @@
+(ns metabase.pulse.markdown
+  (:require [clojure.edn :as edn]
+            [clojure.java.io :as io]
+            [clojure.string :as str]
+            [metabase.public-settings :as public-settings])
+  (:import [com.vladsch.flexmark.ast AutoLink BlockQuote BulletList BulletListItem Code Emphasis FencedCodeBlock
+            HardLineBreak Heading HtmlBlock HtmlCommentBlock HtmlEntity HtmlInline HtmlInlineBase HtmlInlineComment
+            HtmlInnerBlockComment Image ImageRef IndentedCodeBlock Link LinkRef MailLink OrderedList OrderedListItem
+            Paragraph Reference SoftLineBreak StrongEmphasis Text ThematicBreak]
+           com.vladsch.flexmark.parser.Parser
+           [com.vladsch.flexmark.util.ast Document Node]
+           java.net.URI))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                              Markdown parsing                                                  |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(def ^:private parser
+  "An instance of a Flexmark parser"
+  (.build (Parser/builder)))
+
+(def ^:private node-to-tag-mapping
+  "Mappings from Flexmark AST nodes to keyword tags"
+  {Document              :document
+   Paragraph             :paragraph
+   ThematicBreak         :horizontal-line
+   HardLineBreak         :hard-line-break
+   SoftLineBreak         :soft-line-break
+   Heading               :heading
+   StrongEmphasis        :bold
+   Emphasis              :italic
+   OrderedList           :ordered-list
+   BulletList            :unordered-list
+   OrderedListItem       :ordered-list-item
+   BulletListItem        :unordered-list-item
+   Code                  :code
+   FencedCodeBlock       :codeblock
+   IndentedCodeBlock     :codeblock
+   BlockQuote            :blockquote
+   Link                  :link
+   Reference             :reference
+   LinkRef               :link-ref
+   ImageRef              :image-ref
+   Image                 :image
+   AutoLink              :auto-link
+   MailLink              :mail-link
+   HtmlEntity            :html-entity
+   HtmlBlock             :html-block
+   HtmlInline            :html-inline
+   HtmlCommentBlock      :html-comment-block
+   HtmlInlineBase        :html-inline-base
+   HtmlInlineComment     :html-inline-comment
+   HtmlInnerBlockComment :html-inner-block-comment})
+
+(defn- node-to-tag
+  [node]
+  (node-to-tag-mapping (type node)))
+
+(defprotocol ^:private ASTNode
+  (to-clojure [this _]))
+
+(defn- convert-children [node source]
+  (map #(to-clojure % source) (.getChildren ^Node node)))
+
+(extend-protocol ASTNode
+  Node
+  (to-clojure [this source]
+    {:tag     (node-to-tag this)
```

Comment:
one thought, `node-to-tag` and the large map of `node-to-tag-mapping` doesn't seem necessary if we are doing everything inline like this.
```clojure
IndentedCodeBlock
(to-clojure [this _]
  {:tag :codeblock
    :attrs ...})
```

Reply:
Most of the node types don't actually have a custom implementation of `to-clojure` (none of the ones above `FencedCodeBlock` in `node-to-tag-mappings` do). They just hit the `Node` implementation.

We could do everything inline, but that would require a lot of repeated implementations of `to-clojure` with the same structure. I also like how `node-to-tag-mapping` is a concise list of all the node types, which is a nice reference.

Ultimately I could go either way, though.

Reply:
That's a good reason. Good call. It's fine as it is then. I thought it was all repeated for no benefit but you are correct.